usr/src/uts/armv8/os/startup.c

```diff
@@ -1233,6 +1244,11 @@ startup_modules(void)
 	 */
 	seg_init();
 
+	/*
+	 * Set up the interrupt controller for the primary CPU
+	 */
+	startup_gic();
+
```

Comment:
This is why I'm confused about what needs `setlvlx` so soon.  We end up with your changes initializing it fully just after kmem and the vm are up.  What is dealing in interrupts so soon?

Reply:
See my comments above - I'll try to track down exactly what's calling it by looking at the register dumps.

Reply:
On a quick inspection, I think this is due to the locking around prom_printf (lock_set_spl with XC_HI_PIL - hits splr, which passes the new priority through to setlvlx, return path of prom_printf calls splx, forwards to do_splx, which calls setlvlx).

Reply:
So, the locking in `prom_printf` comes from the original arm code from Hayashi, though I changed it to use `lock_set_spl` rather than homebrew it.  The other `prom_printf` implementations are unlocked.  If it makes this cleaner -- and if it works -- we should just unlock the ARM `prom_printf` to match the other platforms.

Reply:
I've never really understood why the locking was needed - the prom printing interfaces shouldn't be used past configure time, should they?

Reply:
They shouldn't, but they used to be in hayashi's original code, and I wasn't confident enough to drop it at the time.